File: src/numpy_wrappers/kernel.cpp
/* ---------------------------------------------------------------------
**
** Copyright (C) 2019 Xiaoyu Wei
**
** Permission is hereby granted, free of charge, to any person obtaining a copy
** of this software and associated documentation files (the "Software"), to deal
** in the Software without restriction, including without limitation the rights
** to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
** copies of the Software, and to permit persons to whom the Software is
** furnished to do so, subject to the following conditions:
** 
** The above copyright notice and this permission notice shall be included in
** all copies or substantial portions of the Software.
** 
** THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
** IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
** FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
** AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
** LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
** OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
** THE SOFTWARE.
**
** -------------------------------------------------------------------*/

namespace pypvfmm{

  enum KernelKind {
    LaplacePotential = 1,
    LaplaceGradient  = 3,

    StokesVelocity   = 5,
    StokesPressure   = 7,
    StokesStress     = 9,
    StokesVelGrad    = 11,

    BiotSavartPotential = 13,

    HelmholtzPotential  = 15,
  };

  const std::unordered_map<std::string, KernelKind>
    kernel_map{
      {"LaplaceKernel, potential", KernelKind::LaplacePotential},
      {"LapKnl3D",                 KernelKind::LaplacePotential},
      {"LaplaceKernel, gradient",  KernelKind::LaplaceGradient},

      {"StokesKernel, velocity", KernelKind::StokesVelocity},
      {"StokesKernel, pressure", KernelKind::StokesPressure},
      {"StokesKernel, stress",   KernelKind::StokesStress},
      {"StokesKernel, vel_grad", KernelKind::StokesVelGrad},

      {"BiotSavartKernel, potential", KernelKind::BiotSavartPotential},

      {"HelmholtzKernel, potential", KernelKind::HelmholtzPotential},
      {"HelmKnl3D(k)",               KernelKind::HelmholtzPotential},
    };

  template <class T>
    const pvfmm::Kernel<T> get_kernel(const std::string &kernel_desc) {
      auto query = kernel_map.find(kernel_desc);
      if (query == kernel_map.end()) {
        throw std::runtime_error("Invalid kernel_desc: " + kernel_desc);
      }
      auto kind = query->second;

      switch(kind) {
        case KernelKind::LaplacePotential:
          return pvfmm::LaplaceKernel<T>::potential();
        case KernelKind::LaplaceGradient:
          return pvfmm::LaplaceKernel<T>::gradient();

        case KernelKind::StokesVelocity:
          return pvfmm::StokesKernel<T>::velocity();
        case KernelKind::StokesPressure:
          return pvfmm::StokesKernel<T>::pressure();
        case KernelKind::StokesStress:
          return pvfmm::StokesKernel<T>::stress();
        case KernelKind::StokesVelGrad:
          return pvfmm::StokesKernel<T>::vel_grad();

        case KernelKind::BiotSavartPotential:
          return pvfmm::BiotSavartKernel<T>::potential();

        case KernelKind::HelmholtzPotential:
          return pvfmm::HelmholtzKernel<T>::potential();
      }
    }

} // end of namespace pypvfmm

File: pypvfmm/kernel.py
__copyright__ = "Copyright (C) 2019 Xiaoyu Wei"

__license__ = """
Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
THE SOFTWARE.
"""

from functools import partialmethod
import numpy as np


class KernelBase():
    def __init__(self, dtype=np.float64):
        assert dtype in [np.float32, np.float64]
        self.dtype = dtype

        if dtype == np.float32:
            self.dtype_str = 'float'
        elif dtype == np.float64:
            self.dtype_str = 'double'


class LaplaceKernel(KernelBase):
    __kernels__ = ['potential', 'gradient']


class StokesKernel(KernelBase):
    __kernels__ = ['velocity', 'pressure', 'stress', 'vel_grad']


class BiotSavartKernel(KernelBase):
    __kernels__ = ['potential']


class HelmholtzKernel(KernelBase):
    __kernels__ = ['potential']


def add_kernels(kernel_container):
    """Add kernel bindings.
    """
    def kernel(self, kernel_name):
        return "%s, %s" % (self.__class__.__name__, kernel_name)

    for kernel_name in kernel_container.__kernels__:
        named_kernel = partialmethod(kernel, kernel_name=kernel_name)
        setattr(kernel_container, kernel_name, named_kernel)


def process_sumpy_kernel(kernel):
    """Parse sumpy kernels to feed to pvfmm.
    """
    if not kernel.dim == 3:
        raise ValueError("PvFMM only supports 3D kernels.")

    from sumpy.kernel import DerivativeBase, LaplaceKernel
    if isinstance(kernel, DerivativeBase):
        if not isinstance(kernel.inner_kernel, LaplaceKernel):
            raise ValueError("PvFMM does not support computing derivatives of %s."
                             % str(kernel.inner_kernel))
        return str(kernel.inner_kernel) + ', gradient'

    return str(kernel)


add_kernels(LaplaceKernel)
add_kernels(StokesKernel)
add_kernels(BiotSavartKernel)
add_kernels(HelmholtzKernel)
